### PR TITLE
feat: dist row inserter

### PR DESCRIPTION
**Pull request by fengjiachun**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Distributed row inserter

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/frontend/src/instance/distributed/row_inserter.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                    table_route
+                        .find_region_leader(region_number)
+                        .context(FindDatanodeSnafu {
+                            region: region_number,
+                        })?;
+                inserts
+                    .entry(peer.clone())
+                    .or_insert_with(|| RowInsertRequests {
+                        inserts: Vec::new(),
+                    })
```
Comment (evenyag) [inline src/frontend/src/instance/distributed/row_inserter.rs:104]: question: Does `RowInsertRequests` implement Default?
```suggestion
                    .or_default()
```

Or
```rust
                    .or_insert_with(RowInsertRequests::default)
```

File: src/partition/src/row_splitter.rs [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+impl RowSplitter {
+    pub fn new(partition_rule: PartitionRuleRef) -> Self {
+        Self { partition_rule }
+    }
+
+    pub fn split(&self, req: RowInsertRequest) -> Result<RowInsertRequestSplits> {
+        // No partition
+        let partition_columns = self.partition_rule.partition_columns();
+        if partition_columns.is_empty() {
+            return Ok(iter::once((0, req)).collect());
```
Comment (evenyag) [inline src/partition/src/row_splitter.rs:41]: non-blocking: You could also use from.
```suggestion
            return Ok(HashMap::from([(0, req)]));
```

**Review by evenyag — APPROVED** [review]

File: src/partition/src/row_splitter.rs [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+            Ok(val)
+        }
+
+        fn find_regions_by_exprs(&self, _: &[PartitionExpr]) -> Result<Vec<RegionNumber>> {
+            unimplemented!()
+        }
+    }
+
+    #[derive(Debug, Serialize, Deserialize)]
+    struct EmptyPartitionRule;
```
Comment (evenyag) [inline src/partition/src/row_splitter.rs:251]: non-blocking: Could we reuse structs like `EmptyPartitionRule`, `MockMissedColPartitionRule`and `MockPartitionRule`  in tests?

**Review by MichaelScofield — APPROVED** [review]

File: src/frontend/src/instance/distributed/row_inserter.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+            common_runtime::spawn_write(async move {
+                let client = datanode_clients.get_client(&peer).await;
+                let database = Database::new(&catalog, &schema, client);
+                database
+                    .row_insert(inserts)
+                    .await
+                    .context(RequestDatanodeSnafu)
+            })
+        }))
```
[the remark this answers is not in the document]
Comment (fengjiachun) [reply src/frontend/src/instance/distributed/row_inserter.rs]: `datanode_clients.get_client(&peer)` is a  async method

File: src/partition/src/row_splitter.rs (hunk 3 above)
Comment (fengjiachun) [reply src/partition/src/row_splitter.rs:251]: Not over yet, stay tuned for the next PR.

**Pull request merged**